base/multidimensional.jl

```diff
@@ -940,3 +940,178 @@ If `dim` is specified, returns unique regions of the array `itr` along `dim`.
         @nref $N A d->d == dim ? sort!(uniquerows) : (1:size(A, d))
     end
 end
+
+"""
+    groupslices(A, dim)
+
+Returns a vector of integers where each integer element of the returned vector
+is a group number corresponding to the unique slices along dimension `dim` as
+returned from `unique(A, dim)`, where `A` can be a multidimensional array.
+
+Example usage:
+
+If `C = unique(A, dim)`, `ic = groupslices(A, dim)`, and
+rank(A) == rank(C) == 3, then:
+
+if dim == 1
+   all(A .== C[ic,:,:])
+elseif dim == 2
+   all(A .== C[:,ic,:])
+elseif dim == 3
+   all(A .== C[:,:,ic])
+end
```

Comment:
Before I make another commit, I need to work get the documentation strings integrated properly as @tkelman points out, but the changes in this [gist](https://gist.github.com/AndyGreenwell/4c2f1481f5c974eb14f8c8266affe85b) should allow your example to work appropriately.  In one of the changes made in the course of discussing #14142 related to detecting hash collisions, where `ic` being directly assigned the values of `uniquerow` was improper.  While the values of `uniquerow` do contain the indices of unique slices of `A`, they are not the indices corresponding to the unique slices in `C` that allow for reconstruction of `A`.

The changes in this [gist](https://gist.github.com/AndyGreenwell/4c2f1481f5c974eb14f8c8266affe85b) assign the appropriate values to `ic` that allow one to reconstruct `A` by indexing `C` with `ic`, as had been the case in the original `uniqueslices` submission (but had not been handling hash collisions properly).

Thank you for testing.